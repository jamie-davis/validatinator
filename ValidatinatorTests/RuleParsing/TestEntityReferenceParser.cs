using System.Linq;
using TestConsole.OutputFormatting;
using TestConsoleLib;
using TestConsoleLib.Testing;
using Validatinator.RuleParsing;
using Xunit;

namespace ValidatinatorTests.RuleParsing
{
    public class TestEntityReferenceParser
    {
        private static readonly string[] TestStrings =
        {
            "FirstError",
            "cleaninput",
            "a.b",
            "a",
            "match(a.b)"
        };

        [Fact]
        public void ParserExtractsRuleComponents()
        {
            //Arrange
            var buffer = new OutputBuffer()
            {
                BufferWidth = 133
            };
            var output = new Output(buffer);

            //Act
            var results = TestStrings
                .Select(s => new {String = s, Result = EntityReferenceParser.Parse(s)})
                .ToList();

            //Assert
            var report = results.AsReport(rep => rep
                .AddColumn(r => r.String, cc => cc.Heading("Input"))
                .AddColumn(r => r.Result.Valid, cc => { })
                .AddChild(r => new[] {new {r.Result?.Error}}, erep => erep.AddColumn(e => e.Error, cc => { }))
                .AddChild(r => new[] {new {Result = r.Result?.EntityReference?.Describe()}},
                    drep => drep.AddColumn(e => e.Result ?? "NULL", cc => cc.Heading("Result")))
            );

            output.FormatTable(report);
            Approvals.Verify(output.Report);
        }
    }
}