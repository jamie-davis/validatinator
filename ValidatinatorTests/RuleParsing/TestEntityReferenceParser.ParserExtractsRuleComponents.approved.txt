Input      Valid                                                                                                                     
---------- -----                                                                                                                     
FirstError False                                                                                                                     
                                                                                                                                     
    Error                                                                                                                            
    --------------------------------                                                                                                 
    Unable to parse entity reference                                                                                                 
                                                                                                                                     
    Result                                                                                                                           
    ------                                                                                                                           
    NULL                                                                                                                             
                                                                                                                                     
Input      Valid                                                                                                                     
---------- -----                                                                                                                     
cleaninput False                                                                                                                     
                                                                                                                                     
    Error                                                                                                                            
    --------------------------------                                                                                                 
    Unable to parse entity reference                                                                                                 
                                                                                                                                     
    Result                                                                                                                           
    ------                                                                                                                           
    NULL                                                                                                                             
                                                                                                                                     
Input      Valid                                                                                                                     
---------- -----                                                                                                                     
a.b        True                                                                                                                      
                                                                                                                                     
    Error                                                                                                                            
    -----                                                                                                                            
                                                                                                                                     
                                                                                                                                     
    Result                                                                                                                           
    ------                                                                                                                           
    a.b                                                                                                                              
                                                                                                                                     
Input      Valid                                                                                                                     
---------- -----                                                                                                                     
a          False                                                                                                                     
                                                                                                                                     
    Error                                                                                                                            
    --------------------------------                                                                                                 
    Unable to parse entity reference                                                                                                 
                                                                                                                                     
    Result                                                                                                                           
    ------                                                                                                                           
    NULL                                                                                                                             
                                                                                                                                     
Input      Valid                                                                                                                     
---------- -----                                                                                                                     
match(a.b) False                                                                                                                     
                                                                                                                                     
    Error                                                                                                                            
    --------------------------------                                                                                                 
    Unable to parse entity reference                                                                                                 
                                                                                                                                     
    Result                                                                                                                           
    ------                                                                                                                           
    NULL                                                                                                                             
                                                                                                                                     
                                                                                                                                     